File: src/services/notificationsService.ts
import {
  collection,
  addDoc,
  query,
  where,
  getDocs,
  doc,
  updateDoc,
  writeBatch,
  Timestamp,
} from "firebase/firestore";
import { firestore } from "@/lib/firebase";

const NOTIFICATIONS_COLLECTION = "notifications";

export type NotificationType = "offer" | "points" | "reward" | "system";

export interface NotificationData {
  id: string;
  userId: string;
  type: NotificationType;
  title: string;
  message: string;
  isRead: boolean;
  createdAt: Date;
  /** Ícone sugerido: tag | star | gift | check | clock | sparkles */
  icon?: string;
  /** Payload opcional para navegação (ex: offerId, storeId) */
  data?: Record<string, string>;
}

function toDate(value: unknown): Date {
  if (!value) return new Date();
  if (typeof (value as { toDate?: () => Date })?.toDate === "function")
    return (value as { toDate: () => Date }).toDate();
  if (typeof (value as { seconds?: number })?.seconds === "number")
    return new Date((value as { seconds: number }).seconds * 1000);
  return new Date();
}

/**
 * Busca notificações do usuário (mais recentes primeiro).
 * Query sem orderBy para não exigir índice composto; ordena em memória.
 */
export async function getNotifications(
  userId: string,
  limitCount: number = 50
): Promise<NotificationData[]> {
  if (!firestore) return [];
  try {
    const q = query(
      collection(firestore, NOTIFICATIONS_COLLECTION),
      where("userId", "==", userId)
    );
    const snapshot = await getDocs(q);
    const items = snapshot.docs.map((docSnap) => {
      const d = docSnap.data();
      return {
        id: docSnap.id,
        userId: String(d?.userId ?? ""),
        type: (d?.type as NotificationType) ?? "system",
        title: String(d?.title ?? ""),
        message: String(d?.message ?? ""),
        isRead: d?.isRead === true,
        createdAt: toDate(d?.createdAt),
        icon: d?.icon as string | undefined,
        data: d?.data as Record<string, string> | undefined,
      };
    });
    items.sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime());
    return items.slice(0, limitCount);
  } catch (error) {
    console.error("Erro ao buscar notificações:", error);
    return [];
  }
}

/**
 * Marca uma notificação como lida (apenas se pertence ao usuário)
 */
export async function markNotificationAsRead(
  notificationId: string,
  userId: string
): Promise<void> {
  if (!firestore) return;
  const ref = doc(firestore, NOTIFICATIONS_COLLECTION, notificationId);
  await updateDoc(ref, { isRead: true });
}

/**
 * Marca todas as notificações do usuário como lidas
 */
export async function markAllNotificationsAsRead(
  userId: string
): Promise<void> {
  if (!firestore) return;
  const q = query(
    collection(firestore, NOTIFICATIONS_COLLECTION),
    where("userId", "==", userId),
    where("isRead", "==", false)
  );
  const snapshot = await getDocs(q);
  const batch = writeBatch(firestore);
  snapshot.docs.forEach((docSnap) => {
    batch.update(docSnap.ref, { isRead: true });
  });
  if (snapshot.docs.length > 0) await batch.commit();
}

/**
 * Conta notificações não lidas do usuário
 */
export async function getUnreadCount(userId: string): Promise<number> {
  if (!firestore) return 0;
  try {
    const q = query(
      collection(firestore, NOTIFICATIONS_COLLECTION),
      where("userId", "==", userId),
      where("isRead", "==", false)
    );
    const snapshot = await getDocs(q);
    return snapshot.size;
  } catch (error) {
    console.error("Erro ao contar notificações não lidas:", error);
    return 0;
  }
}

/**
 * Cria uma notificação para um usuário.
 * Usado quando o lojista confirma um resgate (merchant panel).
 * Requer que o chamador seja merchant (regra Firestore).
 */
export async function createNotification(params: {
  userId: string;
  type: NotificationType;
  title: string;
  message: string;
  icon?: string;
  data?: Record<string, string>;
}): Promise<string> {
  if (!firestore) throw new Error("Firestore não configurado");
  const ref = collection(firestore, NOTIFICATIONS_COLLECTION);
  const docRef = await addDoc(ref, {
    userId: params.userId,
    type: params.type,
    title: params.title,
    message: params.message,
    isRead: false,
    createdAt: Timestamp.now(),
    icon: params.icon ?? iconForType(params.type),
    data: params.data ?? {},
  });
  return docRef.id;
}

function iconForType(type: NotificationType): string {
  switch (type) {
    case "offer":
      return "tag";
    case "points":
      return "star";
    case "reward":
      return "gift";
    default:
      return "check";
  }
}

File: CartaoFidelidade/Services/NotificationsService.swift
//
//  NotificationsService.swift
//  CartaoFidelidade
//
//  Serviço para notificações do usuário no Firestore
//

import Foundation
import FirebaseFirestore
import FirebaseAuth

enum NotificationType: String {
    case offer = "offer"
    case points = "points"
    case reward = "reward"
    case system = "system"
}

struct FirebaseNotification: Identifiable {
    let id: String
    let userId: String
    let type: NotificationType
    let title: String
    let message: String
    let isRead: Bool
    let createdAt: Date
    let icon: String?
    let data: [String: String]?
    
    var sfSymbol: String {
        switch icon ?? "" {
        case "tag": return "tag.fill"
        case "star": return "star.fill"
        case "gift": return "gift.fill"
        case "check": return "checkmark.circle.fill"
        case "clock": return "clock.fill"
        case "sparkles": return "sparkles"
        default:
            switch type {
            case .offer: return "tag.fill"
            case .points: return "star.fill"
            case .reward: return "gift.fill"
            case .system: return "checkmark.circle.fill"
            }
        }
    }
}

class NotificationsService {
    static let shared = NotificationsService()
    private let db = Firestore.firestore()
    private let collectionName = "notifications"
    
    private init() {}
    
    /// Busca notificações do usuário (mais recentes primeiro)
    func getNotifications(userId: String, limitCount: Int = 50) async throws -> [FirebaseNotification] {
        let snapshot = try await db.collection(collectionName)
            .whereField("userId", isEqualTo: userId)
            .getDocuments()
        
        var items = snapshot.documents.compactMap { doc -> FirebaseNotification? in
            parseNotification(docId: doc.documentID, data: doc.data())
        }
        items.sort { $0.createdAt > $1.createdAt }
        return Array(items.prefix(limitCount))
    }
    
    /// Marca uma notificação como lida
    func markAsRead(notificationId: String, userId: String) async throws {
        let ref = db.collection(collectionName).document(notificationId)
        try await ref.updateData(["isRead": true])
    }
    
    /// Marca todas as notificações do usuário como lidas
    func markAllAsRead(userId: String) async throws {
        let snapshot = try await db.collection(collectionName)
            .whereField("userId", isEqualTo: userId)
            .whereField("isRead", isEqualTo: false)
            .getDocuments()
        
        let batch = db.batch()
        for doc in snapshot.documents {
            batch.updateData(["isRead": true], forDocument: doc.reference)
        }
        if !snapshot.documents.isEmpty {
            try await batch.commit()
        }
    }
    
    /// Conta notificações não lidas
    func getUnreadCount(userId: String) async throws -> Int {
        let snapshot = try await db.collection(collectionName)
            .whereField("userId", isEqualTo: userId)
            .whereField("isRead", isEqualTo: false)
            .getDocuments()
        return snapshot.documents.count
    }
    
    /// Cria uma notificação para um usuário (chamado pelo lojista ao confirmar resgate)
    func createNotification(
        userId: String,
        type: NotificationType,
        title: String,
        message: String,
        icon: String? = nil,
        data: [String: String]? = nil
    ) async throws -> String {
        let iconValue = icon ?? iconForType(type)
        let docRef = try await db.collection(collectionName).addDocument(data: [
            "userId": userId,
            "type": type.rawValue,
            "title": title,
            "message": message,
            "isRead": false,
            "createdAt": Timestamp(),
            "icon": iconValue,
            "data": data ?? [:]
        ])
        return docRef.documentID
    }
    
    private func iconForType(_ type: NotificationType) -> String {
        switch type {
        case .offer: return "tag"
        case .points: return "star"
        case .reward: return "gift"
        case .system: return "check"
        }
    }
    
    private func parseNotification(docId: String, data: [String: Any]) -> FirebaseNotification? {
        guard let createdAt = dateFromFirestore(data["createdAt"]) else { return nil }
        let typeRaw = (data["type"] as? String) ?? "system"
        let type = NotificationType(rawValue: typeRaw) ?? .system
        let dataDict = data["data"] as? [String: String]
        return FirebaseNotification(
            id: docId,
            userId: (data["userId"] as? String) ?? "",
            type: type,
            title: (data["title"] as? String) ?? "",
            message: (data["message"] as? String) ?? "",
            isRead: (data["isRead"] as? Bool) ?? false,
            createdAt: createdAt,
            icon: data["icon"] as? String,
            data: dataDict
        )
    }
    
    private func dateFromFirestore(_ value: Any?) -> Date? {
        guard let value = value else { return nil }
        if let ts = value as? Timestamp { return ts.dateValue() }
        if let dict = value as? [String: Any], let seconds = dict["seconds"] as? Int64 {
            return Date(timeIntervalSince1970: TimeInterval(seconds))
        }
        return nil
    }
}
